File: lab22_agrupamento_kmeans.R
# Title     : AGRUPAMENTOS -> USANDO K-MEANS
# Objective : Conhecer como a linguagem R pode nos apoiar nos processos de classificação -> explorando nesta prática o algoritmoK-MEANS
# Created by: accol
# Created on: 08/06/2020

# Nesta prática faremos uso do conjunto de dados iris -> nativo no R
# Esta base -> já possui uma classe (Species) bem definida o que a torna ideal para nossos estudos

cat('\nConhecendo nossa base de dados\n', dim(iris))
cat('\nConhecendo nossa base de dados -> olhando para os atributos\n')
head(iris)
cat('\nConhecendo como as classes estão divididas\n')
summary(iris)
# Podemos observar que nossa base possui exatamente 50 regisgtros para cada classe +=> setosa = 50; versicolor = 50 e virginica = 50
# Praticando => uso do K-MEANS -> observe que não passamos a classe para o agrupador -> no caso aqui sabemos que se trata de 3 grupos - depois discutiremos como escolher o melhor conjunto de grupos
cluster = kmeans(iris[1:4], centers = 3)
cat('\nVamos observar nosso cluster\n')
cluster
cat('\nVamos olhar apenas para os Clusters gerados\n')
cluster$cluster
cluster$betweenss
# Você irá observar que o algoritmo não atribui nomes aos clusters e sim números, neste caso números 1, 2 e 3
# Comparando com o real
table(iris$Species, cluster$cluster)
# Desta tabela podemos observar que nosso algoritmo classificou todas as setosa corretamente, foi muito bem na classificação de versicolor e errou mais em virginica
# Vamos agora visualizar graficamente nossa classificação -> usaremos o plot padrão do R. Neste gráfico estaremos plotando as colunas de um a quatro da nossa base iris e usando como cores -> col = cluster que ele gerou => acompanhe
plot(iris[,1:4], col=cluster$cluster)

File: lab22_agrupamento_kmeans.py
# Title     : AGRUPAMENTOS -> USANDO K-MEANS
# Objective : Conhecer como a linguagem Python pode nos apoiar nos processos de classificação -> explorando nesta prática o algoritmoK-MEANS
# Created by: accol
# Created on: 08/06/2020


from sklearn import datasets
import pandas as pd 
import numpy as np 
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt 
from sklearn.cluster import KMeans


iris = datasets.load_iris()
print('\nA base a ser trabalhada é a base iris')
print('\nO tipo de dado da nossa base é: ', type(iris))
# Vamos fazer uma contagem de quantos elementos temos de cada classe
unicos, quantidade = np.unique(iris.target, return_counts=True)
print('\nOs valores de contagem são "unicos e quantidade" respectivamente \n', unicos, quantidade)

# Ambiente preparado vamos agora trabalhar o algoritmo de agrupamento kMeans. Este algoritmo exige um único argumento que é o número de cluster que você deverá informar

cluster = KMeans(n_clusters=3)
cluster.fit(iris.data)
# Para visualizarmos algum resultado precisamos criar uma variável centróide
centroide = cluster.cluster_centers_
print('\nVamos observar os centroides gerador de cada grupo de nosso cluster: \n', centroide)
# Agora vamos criar uma variável previsoes
previsoes = cluster.labels_
unicos2, quantidade2 = np.unique(previsoes, return_counts=True)

# O próximo passo será avaliar a qualidade do nosso previsor utilizando a Matriz de Confusão
resultados = confusion_matrix(iris.target, previsoes)
print(resultados)

# Agora faremos uma inspeção visual -> usaremos um gráfico de dispersão -> que requer dois atributos x e y -> por isso vamos trabalhar apenas com as colunas 0 e 1
plt.scatter(iris.data[previsoes == 0, 0], iris.data[previsoes == 0, 1], c='green', label='Setosa')
plt.scatter(iris.data[previsoes == 1, 0], iris.data[previsoes == 1, 1], c='red', label='Versicolor')
plt.scatter(iris.data[previsoes == 2, 0], iris.data[previsoes == 2, 1], c='blue', label='Virginica')
plt.legend()
plt.show()
